java/src/main/java/com/saucelabs/simplesauce/SauceOptions.java

```diff
@@ -1,134 +1,150 @@
 
-    public SauceOptions withMac(MacVersion macVersion) {
-        operatingSystem = macVersion.label;
-        browserName = "Safari";
-        return this;
+    // w3c Settings
+    @Getter @Setter private String browserName = "chrome";
+    @Getter @Setter private String browserVersion = "latest";
+    @Getter @Setter private String platformName = "Windows 10";
+    @Getter @Setter private String pageLoadStrategy;
+    @Getter @Setter private Boolean acceptInsecureCerts = null;
+    @Getter @Setter private Proxy proxy;
+    @Getter @Setter private Boolean setWindowRect = null;
+    @Getter @Setter private Map<String, Integer> timeouts;
+    @Getter @Setter private Boolean strictFileInteractability = null;
+    @Getter @Setter private String unhandledPromptBehavior;
+
+    // Sauce Settings
+    @Getter @Setter private Boolean avoidProxy = null;
+    @Setter private String build;
+    @Getter @Setter private Boolean capturePerformance = null;
+    @Getter @Setter private String chromedriverVersion;
+    @Getter @Setter private Integer commandTimeout = null;
+    @Getter @Setter private Map customData = null;
+    @Getter @Setter private Boolean extendedDebugging = null;
+    @Getter @Setter private Integer idleTimeout = null;
+    @Getter @Setter private String iedriverVersion;
+    @Getter @Setter private Integer maxDuration = null;
+    @Getter @Setter private String name;
+    @Getter @Setter private String parentTunnel;
+    @Getter @Setter private Map prerun;
+    @Getter @Setter private String prerunUrl;
+    @Getter @Setter private Integer priority = null;
+    @Getter @Setter private String publicRestricted;
+    @Getter @Setter private Boolean recordLogs = null;
+    @Getter @Setter private Boolean recordScreenshots = null;
+    @Getter @Setter private Boolean recordVideo = null;
+    @Getter @Setter private String screenResolution;
+    @Getter @Setter private String seleniumVersion;
+    @Getter @Setter private List tags = null;
+    @Getter @Setter private String timeZone;
+    @Getter @Setter private String tunnelIdentifier;
+    @Getter @Setter private Boolean videoUploadOnPass = null;
+
+    public static <E extends Enum<E>> boolean isInEnum(String value, Class<E> enumClass) {
+        return Arrays.stream(enumClass.getEnumConstants()).anyMatch(e -> e.name().equals(value));
+    }
+
+    public SauceOptions() {
+        this(new MutableCapabilities());
+    }
+
+    public SauceOptions(MutableCapabilities capabilities) {
+        seleniumCapabilities = capabilities;
+        if (capabilities.getCapability("browserName") != null) {
+            browserName = (String) capabilities.getCapability("browserName");
+        }
+    }
+
+    public MutableCapabilities toCapabilities() {
+        MutableCapabilities w3cCapabilities = getSeleniumCapabilities();
+        MutableCapabilities sauceCapabilities = new MutableCapabilities();
+
+        for (Field field : this.getClass().getDeclaredFields()) {
```

Comment:
This seems awfully complicated. The neater thing to do would be to have `setCapability` check the capability name and set it either on the regular capabilities, or on the sauce capabilities. That means you'll end up with something like:

```java
public SauceOptions {
   private MutableCapabilities topLevel = new MutableCapabilities();
   private MutableCapabilities sauce = new MutableCapabilities();

   public SauceOptions() {
     // Tie the two sets of capabilities together
     topLevel.setCapabilities("sauce:options", sauce);
   }

  ...

  private SauceOptions setCapability(String capabilityName, Object value) {
    Objects.requireNonNull(capabilityName, "Name must be set.");
    if ("sauce:options".equals(capabilityName)) {
      throw new WebDriverException("Cowardly refusing to figure out what to do here");
    }
    if (isInEnum(key, Options.SAUCE.class)) {
      sauce.setCapability(capabilityName, value);
    } else {
      topLevel.setCapability(capabilityName, value);
    }
    return this;
  }
}
```

Reply:
Of course, that's if Lombok can generate a custom setter that calls `setCapability`

Reply:
Yeah, if we're using lombok, I wouldn't want to create a custom setter. It's there to remove boilerplate, and as soon as it is no longer boilerplate, then imo it shouldn't be used (for reasons of explicitness which we agree on). :)

I'd rather not write custom accessors for everything, and I think a generic `setCapability` introduces other problems. So, I'm reluctant to start with either of these, but I'll evaluate if I can accomplish the other things I want without it.